pkg/daemon/cleanup/cleanup.go

```diff
@@ -72,19 +83,19 @@ func Deconfigure(cfg *rest.Config, ctx context.Context, configMapNamespace strin
 	return nil
 }
 
-// getServicesToRemove returns a list of services that should be removed as part of the cleanup process
-// returns the merged Data of the latest ConfigMap, and the ConfigMap specified by the node's version annotation
-func getServicesToRemove(ctx context.Context, cli client.Client, node *core.Node, configMapNamespace string) ([]servicescm.Service, error) {
+// getLatestAndVersionCMData attempts to get the latest and the version CM data that will be used as part of the
+// WICD cleanup process
+func getLatestAndVersionCMData(ctx context.Context, cli client.Client,
+	namespace string, node *core.Node) (*servicescm.Data, *servicescm.Data, error) {
```

Comment:
I'm having a hard time understand why this change is being made. Is it because we no longer only need to merge the service list from each CM, but we also need to know the watched environment variables?
If that is the case instead of making this change, I suggest you change the existing function to return a merged servicescm.Data, instead of two separate CM. Because that is what is actually desired by the caller of this function. Doing this will allow you simplify the `removeServices()` and `cleanup.ensureEnvVarsAreRemoved()` functions as well

Reply:
@sebsoto this has to be done because we need the latest and version CM to be retrieved for both service deletion and environment var deletion.  I would have done what you are suggesting in first place but we only need the merged CM data for services removal and not for env Var Removal. In the previous discussions on this PR, we have decided not to use merged configmap data for env var removal as it is not a complete solution. Here, if the vesion CM data is not available we fall-back to using just the latest CM data.

Reply:
> we need the latest and version CM to be retrieved for both service deletion and environment var deletion

Yes, I see that both of them are passed into both `removeServices` and `ensureEnvVarsAreRemoved`

> but we only need the merged CM data for services removal and not for env Var Removal... we have decided not to use merged configmap data for env var removal as it is not a complete solution

Looking at the code within this PR its not clear to me why that is the case. I don't see how services and envVars would differ in the way they should be handled here. Wouldn't it be safer to use a merged list for the same reasons we need to for services? Partially configured instances which are upgraded will potentially have envVars set by previous versions of WMCO which we need to undo.